### ISSUE TITLE
helmBinary value doesn't work with Layering State Files

### ISSUE THREAD

**Issue by triviman** [roboll/helmfile#1679]
if i don't use layers helmfile works fine with value helmBinary, but I split config in different files , helmfile take default config.

foldering:
```
.
├── bases
│   ├── defaults.yaml
│   ├── environments.yaml
│   └── repos.yaml
├── demo
│   └── INT
│       └── values.yaml
├── externaldns
│   └── INT
│       └── values.yaml
├── fluentbit
│   └── INT
│       └── values.yaml
├── helmfile.yaml
├── README.md
├── releases
│   └── releases.yaml
```

helmfile.yaml:
```
helmDefaults:
  atomic: true
  createNamespace: true
  cleanupOnFail: true
  verify: true
  timeout: 300
  wait: true

bases:
  - "bases/repos.yaml"
  - "bases/environments.yaml"
  - "bases/defaults.yaml"

helmfiles:
  - "releases/releases.yaml"
```

defaults.yaml:

```
commonLabels:
  release: R1
helmBinary: "/home/user1/cloud/bin/helm3"
```

helmfile build:

```
#  Source: helmfile.yaml

filepath: helmfile.yaml
helmBinary: helm
environments:
  int:
    values:
    - replicas: 1
  prod:
    values:
    - replicas: 3
bases:
- bases/repos.yaml
- bases/environments.yaml
- bases/defaults.yaml
helmDefaults:
  tillerless: false
  verify: true
  devel: false
  wait: true
  timeout: 300
  recreatePods: false
  force: false
  atomic: true
  cleanupOnFail: true
  createNamespace: true
  skipDeps: false
  tls: false
helmfiles:
- path: releases/releases.yaml
repositories:
- name: company-nonprod
  url: https://<redacted>
commonLabels:
  release: R1
templates: {}
renderedvalues: {}
```

**Comment by mumoshu** [comment]
@triviman Thanks for reporting! This does seem like a bug in https://github.com/roboll/helmfile/blob/4c0987a6180e8850cd06712dac004608501fbd1b/pkg/state/create.go#L132

**Comment by mumoshu** [comment]
@triviman Ah sorry. I got a bit confused initially, but this seems to be working as intended. You can't inherit anything from the parent helmfile to sub-helmfiles. That applies to hemlfile values, selectors, secrets, helmBinary, and basically all of the settings defined in helmfile.yaml.

This is intentinoal to make each sub-helmfile independently runnable.

If you do need to pass helmBinary from the parent to sub-helmfiles, we'd need to allow specifying it under `helmfiles`, like:

```
helmBinary: helm

helmfiles:
- path: "releases/releases.yaml"
  helmBinary: helm
```

Perhaps enhancing the above to allow reusing the parent's `helmBinary` using go template would also be useful?

```
helmBinary: helm

---

helmfiles:
- path: "releases/releases.yaml"
  helmBinary: {{ .HelmBinary }}
```

Thoughts?

**Comment by triviman** [comment]
@mumoshu , ok I will try your first solution (as part of helmfile declaration).

regards @mumoshu

**Comment by mumoshu** [comment]
@triviman Hey! Thanks for confirming. To be clear, both the first and the second solution aren't possible today as neither `helmBinary` under `helmfiles[]` not `{{ .HelmBinary }}` is not yet implemented.

Do you think we need it? Would you need both or just either of these?

**Comment by mumoshu** [comment]
The possible workaround that should work today would be having this in your parent helmfile.yaml:

```
values:
- helmBinary: path/to/helm

---

{{ $helmBinary := .Values.helmBinary }}

helmBinary: {{ $helmBinary }}

helmfiles:
- path: path/to/sub/helmfile.yaml
  values:
  - helmBinary: {{ $helmBinary }}
```

and something like the below in your sub-helmfile:

```
{{ $helmBinary := .Values.helmBinary }}

helmBinary: {{ $helmBinary }}
```

**Comment by morremeyer** [comment]
@mumoshu You wrote
> This is intentional to make each sub-helmfile independently runnable.

While I can understand that and very much like that idea, it makes it hard to follow a DRY principle. I use `helmfiles` to split up my configuration so that it does not get to big, so I have e.g.:

```
helmfiles:
  - 01-namespaces.yaml
  - 05-crds.yaml
  - 10-sealed-secrets.yaml
  - 30-ingress.yaml
  - 40-monitoring.yaml
``` 

I also run this over multiple clusters, so I’d ideally want to specify

```
helmDefaults:
  kubeContext: home
  wait: true
  createNamespace: false
```

in the same file instead of in every included file. Is there another option to achieve this that I overlooked so far?

**Comment by mumoshu** [comment]
@morremeyer Can you just use `readFile` or `bases` to reuse the structure?

```
{{ readFile "defaults.yaml }}

--

releases:
...
```

```
{{ tpl (readFile "base.yaml.gotmpl") . }}

--

releases:
...
```

```
bases:
- base.yaml

---

releases:
...
```

You need to repeat this in every sub-helmfile.

Be sure not to try inheriting or overwriting sub-helmfile from the parent. That sacrifices maintainability for DRY, which I don't really like to encourage anyone to do.

**Comment by morremeyer** [comment]
I’ll check out using `bases` tomorrow, thank you!

**Comment by mumoshu** [comment]
Thanks for confirming : ) Closing as resolved.
